### ISSUE TITLE
Get rid of staging/ (?)

### ISSUE THREAD

**Issue by dims** [kubernetes/kubernetes#80339]
staging/ is counter to how go modules work. Are there alternatives to what we do today? How about

- git submodules?
- Just move repos out fully and vendor them back in?

What else?

**Comment by dims** [comment]
cc @ibuildthecloud @BenTheElder

**Comment by ibuildthecloud** [comment]
Fundamentally I don't think there is anything wrong with the idea of "submodules" in one git repo.  The conflict is that staging exists in one location and then is expected to be consumed externally from another location (a different git repo).  It's the dual location that messes everything up.  The staging approach has consistently complicated interaction with k8s code externally.

It would be far easier if all of staging just moved to some folder like github.com/kubernetes/kubernetes/foo/client-go, github.com/kubernetes/kubernetes/foo/apiserver, etc and then the expected imports would be k8s.io/kubernetes/foo/client-go.  I know architecturally there is an idea that all of staging should be separate modules with independent lifecycle but they aren't treated that way today and I don't see that they really should be.  It's far, far easier if all these modules followed the same release train and had the same version number.  So I know client-go 1.15.1 is for k8s 1.15.1, for example.  Today this is sort of done with the `kubernetes-X` tags, but unfortunately those aren't semver like so go modules won't respect them.

At this point I'd really request that Kubernetes drop the idea of move staging/* to separate repos and just keep everything in well maintained packages.  I'm sure it would easy enough to write build validation scripts that ensure that nobody introduced "cross module dependencies" that shouldn't exist so as to keep their architectural integrity.

But please, think of the kids, drop staging.

**Comment by BenTheElder** [comment]
staging/ -> cc @sttts @liggitt @nikhita 

I'm not going to argue for or against staging in this comment, but I think a few points should be clarified....

----

> The staging approach has consistently complicated interaction with k8s code externally.

To clarify which "k8s code" are you talking about? The staging libraries / client-go, or packages in this repo that are not in staging?

> go. I know architecturally there is an idea that all of staging should be separate modules with independent lifecycle but they aren't treated that way today and I don't see that they really should be.

AFAIK the reason for staging was not so much seperate module lifecycle as being able to import those things _without_ vendoring the main kubernetes repo (in fact, the staging packages have entirely coupled lifecycles??) 

Importing non-staged packages was / is not really supported, so they needed to be somewhere importable. Staging allows them to be somewhere importable while being developed here with the rest of the core.

> So I know client-go 1.15.1 is for k8s 1.15.1, for example. Today this is sort of done with the kubernetes-X tags, but unfortunately those aren't semver like so go modules won't respect them.

Semver is sort of an issue on it's own, the main repo / Kubernetes is not really semver either. That's somewhat orthogonal to where the packages are published.

> I'm sure it would easy enough to write build validation scripts that ensure that nobody introduced "cross module dependencies" that shouldn't exist so as to keep their architectural integrity.

[import-boss](https://github.com/kubernetes/kubernetes/blob/196047fc280448fadbefe07540065209bf483fa0/staging/src/k8s.io/code-generator/cmd/import-boss/main.go) exists and is a thing used in this repo when we need to ensure that. Again, I don't think staging was originally about architectural integrity, it's about which things are and are not exported / importable.

**Comment by BenTheElder** [comment]
Additionally, a dozen or so packages that are actually decoupled are in another repo without the staging mechanism today at [k8s.io/utils](https://github.com/kubernetes/utils)  (see also: [k8s.io/klog](https://github.com/kubernetes/klog), [sigs.k8s.io/yaml](https://github.com/kubernetes-sigs/yaml)).

**Comment by smarterclayton** [comment]
Staging primarily exists as a way to efficiently refactor and evolve the internal codebase of kubernetes as a unit in the monorepo while still enabling external publishing of useful libraries for the ecosystem.  Making it easy to vendor and fork k/k components has never been a primary goal, although plenty of people have worked to keep it possible for downstreams to patch / alter code.  We have optimized instead for cleanup within the core components while making it still possible for subcomponents to be split out.

Finishing the task of cutting meaningless dependencies between external components like e2e, kube-proxy, Kubelet, and kcm is well within reach, and various other organizational issues also have significant benefits to project health.  Moving away from staging doesn’t seem to offer significant benefits to all but a few consumers today.

**Comment by nikhita** [comment]
>  So I know client-go 1.15.1 is for k8s 1.15.1, for example. Today this is sort of done with the kubernetes-X tags, but unfortunately those aren't semver like so go modules won't respect them.

Experimenting in kubernetes/publishing-bot#196 to start respecting semver and figure out how disruptive versioned imports could be.

**Comment by dims** [comment]
@smarterclayton well said. To be fair, one issue i do have is that we don't have anyone who has graduated from `staging/`. When are we likely to have the first repo move out? 

what we have today in staging/ has worked well to isolate components, but the resistance seems to be "oh if we move it out, then it will be hard to get anything done". Is that a fair statement?

**Comment by smarterclayton** [comment]
If no one had to vendor k/k (because staging move was completed), then there would be no issue, since everyone would just take from the published.  Each time we get a component into staging more people can vendor that component easily.

**Comment by mattfarina** [comment]
A smattering of a few thoughts...

1. What would it take to remove something from staging and instead import it to k/k?
2. There are also git [subtrees](https://git-scm.com/book/en/v1/Git-Tools-Subtree-Merging) as an alternative to submodules.
3. [multi-module repositories](https://github.com/golang/go/wiki/Modules#faqs--multi-module-repositories) are a thing. What happens if something in staging has a go.mod and it's in the standalone repo? I've not looked at this before.

**Comment by sttts** [comment]
Three points of view on staging, from different groups of consumers of k/k:

(a) Staging is under control for in-tree work. 
(b) Staging is awful for people vendoring k/k (how many are that other than OpenShift, k3s?). 
(c) Staging is invisible to people just using our libraries.

(b) is a small group and we should not optimizie primarily for them IMO. Instead, following @smarterclayton 's suggestion: 

* What are the reason that people are in (b) and not in (c) ? What can we do to make them move to (c) ?

With (b), we have another problem: I claim that code in the `pkg/` or `test/` directory in k/k is not suited for public consumption, two magnitues less than our published repos. We don't want to have people access those, and if they do it is good thing that this kind of vendoring is hard.

Splitting up the monorepo is a completely different challenge. Git subtree, submodules have all been discussed at length before. They don't solve **the core problem**:

* Lots of deeply interrelated code without clear bounderies which we can (semantically) version and declare compatibility on. 

Before we have not solved that, moving away from a monorepo will be harmful. We are making progress, but it is a long and painful path.

**Comment by lavalamp** [comment]
> I know architecturally there is an idea that all of staging should be separate modules with independent lifecycle but they aren't treated that way today and I don't see that they really should be. It's far, far easier if all these modules followed the same release train and had the same version number.

@ibuildthecloud  This is how life started out, and all of these things developed inappropriate cross-links to other packages that made it impossible to vendor just a little bit of Kubernetes or e.g. not use our weird build system. Most people strongly do not want to vendor all of our code to use the client.

I agree with @sttts above; vendoring / importing k/k is not supported: we do not try to make it easy or possible and I would explicitly discourage people from attempting it.

IMO the only way to "fix" staging is to just proceed with the original exit plan, which is to delete an item from staging and reverse the import direction so that the external repo is the canonical source. This would take like an hour to actually do, but would cause short-term havoc while people adjusted to the thought of submitting code and then submitting the PR that vendors and uses it.

We got *into* the staging mess because my team had a pressing need to make it remotely possible to consume the client. Exiting the staging mess is important for the community, but not for any particular group--the pain is diffuse, not concentrated. The short-term havoc I mentioned above is way more salient and therefore prevents action.

TL;DR; this is not a technical problem.

**Comment by pwittrock** [comment]
> To be fair, one issue i do have is that we don't have anyone who has graduated from staging/. When are we likely to have the first repo move out?

@dims  kubectl is going through this process now.  *out* could mean a couple of different things , but I expect 6-9 months at the current rate of contributions.  @seans3 has been leading this, and we have had a number of other contributors from the community.

**Comment by dims** [comment]
@pwittrock ack, to be noted that `kubectl` just got into staging/ so no issues with that one for sure! and thanks to you all to drive that effort to get it out as well.

**Comment by mattfarina** [comment]
> I claim that code in the pkg/ or test/ directory in k/k is not suited for public consumption, two magnitues less than our published repos. We don't want to have people access those, and if they do it is good thing that this kind of vendoring is hard.

I would like to point out that people are using packages in `pkg` today because there is no good alternative for some things. To avoid this they would have to copy the code out and keep up with the changes as they land in k/k. I don't know anyone that likes doing this but the alternative isn't appealing enough that they have done it.

I'm not suggesting we make it easier for folks. Just reminding folks of the state of things.

> IMO the only way to "fix" staging is to just proceed with the original exit plan, which is to delete an item from staging and reverse the import direction so that the external repo is the canonical source. This would take like an hour to actually do, but would cause short-term havoc while people adjusted to the thought of submitting code and then submitting the PR that vendors and uses it.

+3000. The hardest is the first one. Which is the easiest one to do and how do we proceed?

**Comment by sttts** [comment]
> I would like to point out that people are using packages in `pkg` today because there is no good alternative for some things. To avoid this they would have to copy the code out and keep up with the changes as they land in k/k. I don't know anyone that likes doing this but the alternative isn't appealing enough that they have done it.

Would be interesting to analyse what these packages in `pkg` are.

**Comment by dims** [comment]
+100 @sttts that's how we ended up staging the cri stuff. k3s is a unique use case but yes we need to find out which ones are useful especially for controllers(?) and operators(?)

**Comment by smarterclayton** [comment]
I'm not even sure it's *that* much more work.  We could in a single release
move most of the remaining components to staging, including kube-apiserver
(which would be everything that is left).  Kubelet, kube-proxy, Kcm, and
scheduler would go fast.  Tests would go fast once someone decides where to
put kubemark (which is bundled in e2e and depends on kubelet).  hyperkube
might be a bit uglier, but it's not *that* bad.

I'd almost say it's worth it for 1.17 just so we stop having the circular
arguments.

**Comment by ibuildthecloud** [comment]
Just to clarify a one thing, k3s is a very unique project and it's needs should ***not*** be considered.  The frustration with importing k8s code is for just writing controllers, which have to import code that originates from staging/.  If you have any code that depends on k8s.io/api and/or k8s.io/client-go and try to use modules, and your dependencies also have dependencies on k8s.io (which is common for controllers), you are in a hell of a mess.

The basic problem is that api, client-go, api-machinery, code-generator, apiexternstions-apiserver are often tightly coupled.  This is basically a side effect that they are managed/developed as one unit then copied to k8s.io/* as separate units.  So you need to use the proper combination of those packages.  The easiest thing is to just lock them all to a `kubernetes-1.x.x` tag but this is not so easy with go modules.  Because it's not semver looking so go modules will turn the pretty tag into a git commit hash.  So it's really hard to know if you have the right collection of packages. (Not to mention that all k8s.io/* projects use go.mod replace statements which are ignored when used as a dependency).

I'm sure you'll just stick with the current staging plan.  What ever you do, I would just really like it worked now, not 2 years from now.  I personally would not advocate for multiple git repos because the management overhead is huge.  But Red Hat and Google have a crap ton of engineers so maybe you all can handle it.

What I'd personally prefer, as a writer of a ton of controllers, is that if I need to develop a controller I just depend on one thing k/k, not 9+ k8s.io packages that have breaking inter-dependencies.

One small thing that I think would really help is that instead of tagging k8s.io packages that come from staging as kubernetes-1.x.x, instead tag them as v1.x.x so that at least reading go.mod one could quickly infer that the right set of dependencies exist.

**Comment by ibuildthecloud** [comment]
Oh another thing, reading through this thread I'm thoroughly confused on the purpose of staging.  So do what you will, I bow out, I don't get it.

But, just to be clear, using k8s dependencies through go modules is really, really confusing.  The typical use case would be like "I want to write a controller against k8s 1.14.0".  So I end up with a go.mod like

```
	k8s.io/api v0.0.0-20190409021203-6e4e0e4f393b
	k8s.io/apiextensions-apiserver v0.0.0-20190409022649-727a075fdec8
	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
	k8s.io/code-generator v0.0.0-20190311093542-50b561225d70
	k8s.io/gengo v0.0.0-20190327210449-e17681d19d3a
	k8s.io/kube-openapi v0.0.0-20190502190224-411b2483e503 // indirect
	k8s.io/utils v0.0.0-20190506122338-8fab8cb257d5 // indirect
	sigs.k8s.io/yaml v1.1.0 // indirect
```

It's hard to know what any of that means and if any of those are really the proper dependencies.  So typically you just go with "it compiles" so it must be good.  It just sucks when it doesn't compile and you're lost.  And this is for writing a "sample controller" style controller.  Now if you try to use controller-runtime you look at controller-runtime go.mod and it has 

```
        k8s.io/api v0.0.0-20190409021203-6e4e0e4f393b
	k8s.io/apiextensions-apiserver v0.0.0-20190409022649-727a075fdec8
	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
	k8s.io/kube-openapi v0.0.0-20180731170545-e3762e86a74c // indirect
	k8s.io/utils v0.0.0-20190506122338-8fab8cb257d5 // indirect
	sigs.k8s.io/testing_frameworks v0.1.1
        sigs.k8s.io/yaml v1.1.0
````

Now imagine if you depend on another project because you want to import their api structs.  Then you end up with another set of random projects that that project depends on.  It's just a big mess.

**Comment by munnerz** [comment]
> So it's really hard to know if you have the right collection of packages. (Not to mention that all k8s.io/* projects use go.mod replace statements which are ignored when used as a dependency).

FWIW, in order to force go modules to use the correct (or as you put it, the 'same' version across all repos), you can use the `replace` directive in your own projects go.mod file, like this: https://github.com/jetstack/cert-manager/blob/d9899e86bec20645c13f3a99e1e48bbf49399d23/go.mod#L125-L139

Whilst these still encode go module's 'reproducible' ref format, you can put branch names, tag names (e.g. kubernetes-1.15.0), commit refs, whatever you want in there and `go mod tidy` will automatically resolve it for you.

In cert-manager I've done this and have very few issues with our k8s.io/* dependencies. I've also used this in order to depend *on* cert-manager from another CRD based project, built using controller-runtime (which itself depends on `k8s.io/client@v11.0`).

> instead tag them as v1.x.x so that at least reading go.mod one could quickly infer that the right set of dependencies exist.

Ultimately, as others have stated, this is caused by k8s itself *not* being produced as a library, so moving all the code under `pkg/` would mean that the go module situation would be just as bad (if not worse), as k8s would tag new revisions of what go modules would define as a 'library' as a new *minor* version, despite there potentially being breaking changes within that version bump.

This is precisely why the currently exported staging/ repositories are *not* tagged with semver compliant tags, because they *aren't* semver compliant. The only way to make them semver compliant would be to tag a new *major* release upon each k8s *minor* release (and potentially even in *patch* releases if some parts of exported code were to be modified within a patch release).

> It's hard to know what any of that means and if any of those are really the proper dependencies. So typically you just go with "it compiles" so it must be good.

I'd *strongly* recommend that you use the `replace` directive to control this, it's worked well for us 😄

> I'd almost say it's worth it for 1.17 just so we stop having the circular

If we could get even just one repo on its way with this it'd be amazing 🙌 could we get some tasks enumerated into a tracking issue, broken down, and then we can start to help out? 😄

**Comment by yue9944882** [comment]
> Would be interesting to analyse what these packages in pkg are.

i've noticed a number of projects importing k/k directly, even nowadays the crowd is getting larger considering those internal ones i presume.. here're a few reasons i can remember of driving folks to persist in importing k/k:
- directly importing internal types. even tho this is getting more difficult in the recent releases b/c we removed client library for internal types (tracked in #70657), but by scanning github, i see other projects converting objects to internal version on purpose to make use of validation helpers i.e. `pkg/apis/<group>/validation`, this mostly happens to those developing third-party controllers.
- importing helper functions working for versioned types. e.g.`pkg/apis/core/v1/helper`. (folks are sometimes leaning to import them instead of copy/pasting the helper fractions.)
- importing controller helper functions for developing controllers. `pkg/controller/...` (this will be resolved as KCM goes out-of-tree)
- importing e2e testing framework to for their own integration test. i.e. `test/...`

**Comment by pohly** [comment]
James Munnelly <notifications@github.com> writes:
> This is precisely why the currently exported staging/ repositories are
> *not* tagged with semver compliant tags, because they *aren't* semver
> compliant. The only way to make them semver compliant would be to tag
> a new *major* release upon each k8s *minor* release

This is indeed the approach currently under investigation, see
https://github.com/kubernetes/kubernetes/issues/72638

> (and potentially
> even in *patch* releases if some parts of exported code were to be
> modified within a patch release).

I hope that this will never happen. But yes, it might :-(

-- 
Best Regards

Patrick Ohly

**Comment by ibuildthecloud** [comment]
I don't feel there is enough understanding of the pains users go through to try to consume k8s libraries in perfectly valid approaches using go modules.

I very much DO NOT want staging libraries to stay in multiple git repos. There are so many packages and it's too much burden on the user on which versions to use. And the nature of go modules is it changes dependencies automatically based on transient dependencies so you quickly get lost.

While @thockin is half joking, his proposal isn't far off from what I think is ideal. What I'd like to see is.

1. All go code moves to a package like internal/
2. All code in staging/ moves to the root of k/k.

So anything in k/k is fair use except internal/ and everything is versioned as one unit.

The only thing is internal/ might be too harsh because go actually disallows importing internal/ and while it's supposed to be unsupported I always hate preventing people to shoot themselves in the foot that specifically want to.

**Comment by deitch** [comment]
@munnerz, you wrote:

> In cert-manager I've done this and have very few issues with our k8s.io/* dependencies. I've also used this in order to depend on cert-manager from another CRD based project, built using controller-runtime (which itself depends on k8s.io/client@v11.0).

Would you consider a canonical guide to safe importing? I think @ibuildthecloud is correct in that the pain in importing something just to build a client, controller, etc is not fully appreciated. Having an official guide and some samples on how to do it safely and cleanly, while not addressing the core issues raised entirely, would be a huge help to those of us doing it under the current structure.

**Comment by pohly** [comment]
@munnerz wrote:
> Whilst these still encode go module's 'reproducible' ref format, you can put branch names, tag names (e.g. kubernetes-1.15.0), commit refs, whatever you want in there and go mod tidy will automatically resolve it for you.

Which I found to be not so useful when it came to reviewing the resulting change. When looking at a modified `go.mod` file after `go mod tidy`, a reviewer only gets to see the git hash. It is possible to see whether those are all consistent, but that isn't obvious and the original intent ("use Kubernetes 1.15.0") is lost.

Not running `go mod tidy` isn't an option either, because it is needed to ensure that the file is really up-to-date.

My conclusion is that Kubernetes urgently needs proper sematic version tags for packages that are supposed to be usable outside of Kubernetes, i.e. https://github.com/kubernetes/kubernetes/issues/72638 needs to be resolved. With `dep` it was kind of okay to have the `kubernetes-vX.Y.Z` tags, but with `go mod` the situation became worse.
